### DISCUSSION TITLE
Proposal: `LiteralInt`

### DISCUSSION THREAD

**Discussion by elenakrittik** [python/typing#1665] [General]
Similar to the `LiteralString` type, `LiteralInt` is an integer that is (or is composed of) constant integer literals.

#### The problem

Identificators (IDs) in many applications are represented as integers, and it is convenient from time to time to just hardcode an entity's ID that you know will not change when the API you use doesn't provide good means to programmatically find the necessary entity. This is the exact use case i have: i want all IDs in a library to be strongly typed and differentiated (so that you can't accidentally misuse, e.g., a user ID when a message ID was expected, as they will be different NewTypes of `int`), but this creates a limitation: i can no longer pass raw `int` literal where i used to do so, so i'll need to find all usages of that pattern and update them, which is prohibitive given the amount of code that already depends on this type-checking properly. I can't use something like `UserID | int` either because that will essentially eliminate the whole point and silently cast NewTypes of other ID types to `int` and therefore pass the type check. `LiteralInt` would solve this.

I'd like to hear others' opinion on this: is it worth it? Maybe some other use cases you have? Any feedback appreciated.

**Comment by Gobot1234** [comment]
I agree it seems like a good idea but I think this'd be much better generalised by Literal[int] working that way you get the other types that work with Literal for free

**Reply by Daverball** [reply]
I think even more interesting would be adding support for slice objects, e.g. `Literal[0:]` for all positive integers or `Literal[0:65535]` for all valid port numbers. Then you could write `LiteralInt` as `Literal[:]`.

Similarly we could benefit from RegEx support for string literals.

**Reply by elenakrittik** [reply]
I agree such a generalisation would be useful, although it would require a little more effort to write up. Also, [we'll have to some up with a good name for it](https://peps.python.org/pep-0675/#rejected-names).

**Reply by Daverball** [reply]
If there is any pushback on reusing `Literal` I'd suggest `LiteralPattern`, this makes sense for both slices and regular expressions.

**Reply by elenakrittik** [reply]
I was thinking of `ArbitraryLiteral`, inspired by `LiteralString` PEP's title ("Arbitrary Literal String Type")

**Reply by TeamSpen210** [reply]
Perhaps `LiteralOf[int]`?

**Reply by jorenham** [reply]
Or perhaps `Just[int]`, for the sake of brevity.

**Reply by jorenham** [reply]
@Daverball A literal slice-type sounds awesome for integers, but only for integers (and perhaps floats): `Literal['spam':]` or `Literal[Color.RED:]` (`Color` is an enum) don't make much sense.

**Reply by jorenham** [reply]
Alternatively, what about naming it `Constant[int]` or `Const[int]`?

**Comment by jorenham** [comment]
This would make it possible to provide correct typeshed annotations for `len()`, which currently upcasts any `Literal[:int]` returned from `__len__()` to an `int`.

**Reply by jorenham** [reply]
See https://github.com/python/typeshed/pull/11830 for a related discussion

**Comment by randolf-scholz** [comment]
Another great use case would be for annotating `matplotlib`'s [`subplot`](https://github.com/matplotlib/matplotlib/blob/80c96ef3cfc464de297a9003e558deb3dc5c643b/lib/matplotlib/pyplot.py#L1553-L1614).

Currently, it seems one has to manually add `squeeze=False` to make the type-checker infer `ndarray`. With `Literal[int]` one could write overloads that automatically infer this case if ncols>1 or nrows>1.
